sw/host/ot_certs/src/template/mod.rs

```diff
@@ -0,0 +1,571 @@
+    pub variables: HashMap<String, VariableType>,
+    /// Certificate specification.
+    pub certificate: Certificate,
+}
+
+/// Certificate specification.
+#[serde_as]
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize)]
+pub struct Certificate {
+    #[serde_as(as = "Value<hjson::BigUint>")]
+    pub serial_number: Value<BigUint>,
+    pub issuer: HashMap<AttributeType, Value<String>>,
+    pub subject: HashMap<AttributeType, Value<String>>,
+    pub subject_public_key_info: SubjectPublicKeyInfo,
+    #[serde_as(as = "Value<hjson::HexString>")]
+    pub authority_key_identifier: Value<Vec<u8>>,
+    #[serde_as(as = "Value<hjson::HexString>")]
+    pub subject_key_identifier: Value<Vec<u8>>,
+    pub model: Option<Value<String>>,
+    pub vendor: Option<Value<String>>,
+    pub version: Option<Value<String>>,
+    #[serde_as(as = "Option<Value<hjson::BigUint>>")]
+    pub svn: Option<Value<BigUint>>,
+    #[serde_as(as = "Option<Value<hjson::BigUint>>")]
+    pub layer: Option<Value<BigUint>>,
+    pub fw_ids: Option<Vec<FirmwareId>>,
+    pub flags: Option<Flags>,
+    pub signature: Signature,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Hash, strum::Display)]
+#[serde(rename_all = "snake_case")]
+pub enum AttributeType {
+    #[serde(alias = "c")]
+    Country,
+    #[serde(alias = "o")]
+    Organization,
+    #[serde(alias = "ou")]
+    OrganizationalUnit,
+    #[serde(alias = "st")]
+    State,
+    #[serde(alias = "cn")]
+    CommonName,
+    #[serde(alias = "sn")]
+    SerialNumber,
+}
+
+/// Value which may either be a variable name or literal.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Value<T> {
+    /// This value will be populated on the device when variables are set.
+    Variable {
+        /// Name of the variable.
+        var: String,
+        /// Optional conversion to apply to the variable.
+        convert: Option<Conversion>,
+    },
+    /// Constant literal that will be set when the certificate is generated.
+    Literal(T),
+}
+
+impl<T> Value<T> {
+    /// Create a variable with the given name. No conversion applied.
+    pub fn variable(name: &str) -> Self {
+        Value::Variable {
+            var: String::from(name),
+            convert: None,
+        }
+    }
+
+    /// Create a variable with the given name and conversion.
+    pub fn convert(var: &str, conversion: Conversion) -> Self {
+        Value::Variable {
+            var: String::from(var),
+            convert: Some(conversion),
+        }
+    }
+
+    /// Create a literal with the given value.
+    pub fn literal(value: impl Into<T>) -> Self {
+        Value::Literal(value.into())
+    }
+
+    /// Return true if the value is a literal
+    pub fn is_literal(&self) -> bool {
+        matches!(self, Self::Literal(_))
+    }
+}
+
+// Manual implementation of the deserializer since error messages for untagged
+// enumerations are really terrible.
+impl<'de, T, U> DeserializeAs<'de, Value<T>> for Value<U>
```

Comment:
Hmm I not sure I follow here, what about the error messages requires a manual implementation of the deserializer here?

Reply:
The error message by serde is `the data does not match an untagged enum`: it won't even tell you what it expects! Part of the issue is also that the various fields have types (e.g. serial number is an integer, vendor is a string) but this is not really obvious and it does affect the parsing. With this, the tool actually tells you what are the possible options which is slightly better. This is a well-known issue of serde that the developers won't fix.

Reply:
ah got it. thanks for clarifying.